nvme_gw_persistence.py

```diff
@@ -0,0 +1,391 @@
+                self._delete_key(key)
+
+    def _restore_subsystems(self, omap_dict, callback):
+        """Restores subsystems from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.SUBSYSTEM_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                req = pb2.subsystem_create_req(
+                    subsystem_nqn=args["subsystem_nqn"],
+                    serial_number=args["serial_number"],
+                )
+                callback(req)
+
+    def add_host(self, subsystem_nqn: str, host_nqn: str, val: str):
+        """Adds a host to the persistent config."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._write_key(key, val)
+
+    def delete_host(self, subsystem_nqn: str, host_nqn: str):
+        """Deletes a host from the persistent config."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._delete_key(key)
+
+    def _restore_hosts(self, omap_dict, callback):
+        """Restore hosts from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.HOST_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                if "allow_any_host" in args:
+                    req = pb2.subsystem_add_host_req(
+                        subsystem_nqn=args["subsystem_nqn"],
+                        allow_any_host=True,
+                    )
+                else:
+                    req = pb2.subsystem_add_host_req(
+                        subsystem_nqn=args["subsystem_nqn"],
+                        host_nqn=args["host_nqn"],
+                    )
+                callback(req)
+
+    def add_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str,
+                     val: str):
+        """Adds a listener to the persistent config."""
+        key = "{}{}_{}_{}".format(self.LISTENER_PREFIX, subsystem_nqn, traddr,
+                                  trsvcid)
+        self._write_key(key, val)
+
+    def delete_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str):
+        """Deletes a listener from the persistent config."""
+        key = "{}{}_{}_{}".format(self.LISTENER_PREFIX, subsystem_nqn, traddr,
+                                  trsvcid)
+        self._delete_key(key)
+
+    def _restore_listeners(self, omap_dict, callback):
+        """Restores listeners from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.LISTENER_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                req = pb2.subsystem_add_listener_req(
+                    nqn=args["nqn"],
+                    trtype=args["trtype"],
+                    adrfam=args["adrfam"],
+                    traddr=args["traddr"],
+                    trsvcid=args["trsvcid"],
+                )
```

Comment:
Since the entire message is serialized and then deserialized, this could probably be shortened (here and in other `_restore_*` functions):
```suggestion
                req = pb2.subsystem_add_listener_req(**args)
```